packages/@aws-cdk/toolkit-lib/lib/api/cloud-assembly/private/prepare-source.ts

```diff
@@ -167,59 +167,45 @@ export class ExecutionEnvironment implements AsyncDisposable {
-    let contextOverflowLocation = null;
+/**
+ * Serializes the given context to a set if environment variables environment variables
+ *
+ * Needs to know the size of the rest of the env because that's necessary to do
+ * an overflow computation on Windows. This function will mutate the given
+ * environment in-place. It should be called as the very last operation on the
+ * environment, because afterwards is might be at the maximum size.
+ *
+ * This *would* have returned an `IAsyncDisposable` but that requires messing
+ * with TypeScript type definitions to use it in aws-cdk, so returning an
+ * explicit cleanup function is easier.
+ */
+export function writeContextToEnv(env: Env, context: Context) {
+  let contextOverflowLocation = null;
 
-    try {
-      const envVariableSizeLimit = os.platform() === 'win32' ? 32760 : 131072;
-      const [smallContext, overflow] = splitBySize(context, spaceAvailableForContext(env, envVariableSizeLimit));
+  // On Windows, all envvars together must fit in a 32k block (<https://devblogs.microsoft.com/oldnewthing/20100203-00>)
+  // On Linux, a single entry may not exceed 131k; but we're treating it as all together because that's safe
+  // and it's a single execution path for both platforms.
+  const envVariableSizeLimit = os.platform() === 'win32' ? 32760 : 131072;
+  const [smallContext, overflow] = splitBySize(context, spaceAvailableForContext(env, envVariableSizeLimit));
 
-      // Store the safe part in the environment variable
-      env[cxapi.CONTEXT_ENV] = JSON.stringify(smallContext);
+  // Store the safe part in the environment variable
+  env[cxapi.CONTEXT_ENV] = JSON.stringify(smallContext);
 
-      // If there was any overflow, write it to a temporary file
-      if (Object.keys(overflow ?? {}).length > 0) {
-        const contextDir = fs.mkdtempSync(path.join(os.tmpdir(), 'cdk-context'));
-        contextOverflowLocation = path.join(contextDir, 'context-overflow.json');
-        fs.writeJSONSync(contextOverflowLocation, overflow);
-        env[cxapi.CONTEXT_OVERFLOW_LOCATION_ENV] = contextOverflowLocation;
-      }
+  // If there was any overflow, write it to a temporary file
+  if (Object.keys(overflow ?? {}).length > 0) {
+    const contextDir = fs.mkdtempSync(path.join(os.tmpdir(), 'cdk-context'));
+    contextOverflowLocation = path.join(contextDir, 'context-overflow.json');
+    fs.writeJSONSync(contextOverflowLocation, overflow);
+    env[cxapi.CONTEXT_OVERFLOW_LOCATION_ENV] = contextOverflowLocation;
+  }
 
-      // call the block code with new environment
-      return await block(env, context);
-    } finally {
-      if (contextOverflowLocation) {
-        fs.removeSync(path.dirname(contextOverflowLocation));
-      }
+  return async () => {
+    if (contextOverflowLocation) {
+      await fs.promises.rm(path.dirname(contextOverflowLocation), { recursive: true, force: true });
     }
-  }
+  };
```

Comment:
Why not implement AsyncDisposable?

Reply:
Led to pains in using it in `aws-cdk` that I didn't want to deal with.